File: apps/api/app/schemas/benchmarks.py
from datetime import datetime
from decimal import Decimal
from uuid import UUID

from pydantic import BaseModel, ConfigDict, Field

from app.schemas.enums import Provider, RunStatus


class BenchmarkCreate(BaseModel):
    name: str = Field(min_length=1)
    description: str = ""
    etalon_ids: list[UUID] = Field(min_length=1)
    skill_id: UUID
    provider: Provider
    model: str = Field(min_length=1)
    judge_skill_id: UUID
    evaluation_mode: str = "layer_1_and_layer_2"
    run_parameters: dict = Field(default_factory=dict)


class BenchmarkRead(BaseModel):
    id: UUID
    name: str
    description: str
    etalon_ids: list[UUID]
    skill_id: UUID
    skill_version: str
    judge_skill_id: UUID
    provider: Provider
    model: str
    status: RunStatus
    started_by_id: UUID
    started_at: datetime | None
    completed_at: datetime | None
    overall_score: Decimal | None
    layer_1_score: Decimal | None
    layer_2_score: Decimal | None
    precision: Decimal | None
    recall: Decimal | None
    f1: Decimal | None
    missed_findings: list | None
    false_positives: list | None
    partial_matches: list | None
    judge_output: dict | None
    report: dict | None
    run_parameters: dict
    error_message: str | None

    model_config = ConfigDict(from_attributes=True)


class BenchmarksListResponse(BaseModel):
    benchmarks: list[BenchmarkRead]

File: apps/web/src/lib/api/benchmarks.ts
import { apiFetch } from "./client";
import type { Provider, RunStatus } from "./documents";

export type BenchmarkRecord = {
  id: string;
  name: string;
  description: string;
  etalon_ids: string[];
  skill_id: string;
  skill_version: string;
  judge_skill_id: string;
  provider: Provider;
  model: string;
  status: RunStatus;
  started_by_id: string;
  started_at: string | null;
  completed_at: string | null;
  overall_score: string | null;
  layer_1_score: string | null;
  layer_2_score: string | null;
  precision: string | null;
  recall: string | null;
  f1: string | null;
  missed_findings: unknown[] | null;
  false_positives: unknown[] | null;
  partial_matches: unknown[] | null;
  judge_output: Record<string, unknown> | null;
  report: Record<string, unknown> | null;
  run_parameters: Record<string, unknown>;
  error_message: string | null;
};

export type BenchmarkCreatePayload = {
  name: string;
  description: string;
  etalon_ids: string[];
  skill_id: string;
  provider: Provider;
  model: string;
  judge_skill_id: string;
  evaluation_mode: string;
  run_parameters: Record<string, unknown>;
};

export async function listBenchmarks(): Promise<{ benchmarks: BenchmarkRecord[] }> {
  return apiFetch<{ benchmarks: BenchmarkRecord[] }>("/benchmarks");
}

export async function createBenchmark(payload: BenchmarkCreatePayload): Promise<BenchmarkRecord> {
  return apiFetch<BenchmarkRecord>("/benchmarks", { method: "POST", body: JSON.stringify(payload) });
}

export async function getBenchmark(benchmarkId: string): Promise<BenchmarkRecord> {
  return apiFetch<BenchmarkRecord>(`/benchmarks/${benchmarkId}`);
}

export async function getBenchmarkReport(benchmarkId: string): Promise<Record<string, unknown>> {
  return apiFetch<Record<string, unknown>>(`/benchmarks/${benchmarkId}/report`);
}

export async function cancelBenchmark(benchmarkId: string): Promise<BenchmarkRecord> {
  return apiFetch<BenchmarkRecord>(`/benchmarks/${benchmarkId}/cancel`, { method: "POST" });
}
